src/stream/stream/last.rs

```diff
@@ -0,0 +1,42 @@
+use std::pin::Pin;
+
+use crate::future::Future;
+use crate::stream::Stream;
+use crate::task::{Context, Poll};
+
+#[doc(hidden)]
+#[allow(missing_debug_implementations)]
+pub struct LastFuture<S, T> {
+    stream: S,
+    last: Option<T>,
+}
+
+impl<S, T> LastFuture<S, T> {
+    pin_utils::unsafe_pinned!(stream: S);
+    pin_utils::unsafe_pinned!(last: Option<T>);
+
+    pub(crate) fn new(stream: S) -> Self {
+        LastFuture { stream, last: None }
+    }
+}
+
+impl<S> Future for LastFuture<S, S::Item>
+where
+    S: Stream + Unpin + Sized,
+    S::Item: Copy,
+{
+    type Output = Option<S::Item>;
+
+    fn poll(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let next = futures_core::ready!(self.as_mut().stream().poll_next(cx));
+
+        match next {
+            Some(new) => {
+                cx.waker().wake_by_ref();
+                *self.as_mut().last() = Some(new);
```

Comment:
This assignment does not compile, how can I get that working. The `MinByFuture` does exactly the same, but obviously I am missing _something_ but can't see it.

Reply:
@yoshuawuyts @stjepang

Reply:
Compile Error:
```
error[E0594]: cannot assign to data in a dereference of `std::pin::Pin<&mut std::option::Option<<S as futures_core::stream::Stream>::Item>>`
  --> src/stream/stream/last.rs:36:17
   |
36 |                 *self.as_mut().last() = Some(new);
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot assign
   |
   = help: trait `DerefMut` is required to modify through a dereference, but it is not implemented for `std::pin::Pin<&mut std::option::Option<<S as futures_core::stream::Stream>::Item>>`

error: aborting due to previous error

error: could not compile `async-std`.
```

Reply:
FYI: If you want to assign to the pinned value, you can use [`Pin::set`](https://doc.rust-lang.org/nightly/core/pin/struct.Pin.html#method.set).

```rust
self.as_mut().last().set(Some(new));
```

Reply:
right, I have to take a proper look at pinning. I haven't used it before.